core/server/src/main/java/alluxio/worker/file/FileDataManager.java

```diff
@@ -292,34 +298,46 @@ public void persistFile(long fileId, List<Long> blockIds) throws IOException {
    * @throws IOException if the folder creation fails
    */
-  private String prepareUfsFilePath(long fileId) throws IOException {
+  private String prepareUfsFilePath(long fileId) throws AlluxioException, IOException {
     String ufsRoot = Configuration.get(PropertyKey.UNDERFS_ADDRESS);
     FileInfo fileInfo = mBlockWorker.getFileInfo(fileId);
     AlluxioURI uri = new AlluxioURI(fileInfo.getPath());
-    String dstPath = PathUtils.concatPath(ufsRoot, fileInfo.getPath());
+    AlluxioURI dstPath = new AlluxioURI(PathUtils.concatPath(ufsRoot, fileInfo.getPath()));
     LOG.info("persist file {} at {}", fileId, dstPath);
     String parentPath = PathUtils.concatPath(ufsRoot, uri.getParent().getPath());
     // creates the parent folder if it does not exist
     if (!mUfs.isDirectory(parentPath)) {
-      final int maxRetry = 10;
-      int numRetry = 0;
-      // TODO(peis): Retry only if we are making progress.
-      // TODO(chaomin): figure out a way to get parent permission in Alluxio namespace.
-      for (; numRetry < maxRetry; numRetry++) {
-        if (mUfs.mkdirs(parentPath) || mUfs.isDirectory(parentPath)) {
-          break;
-        }
-        // The parentPath can be created between the exists check and mkdirs call by other threads.
-        LOG.warn("Failed to create dir: {}, retrying", parentPath);
+      FileSystem fs = FileSystem.Factory.get();
+      // Create ancestor directories from top to the bottom. We cannot use recursive create parents
+      // here because the permission for the ancestors can be different.
+      Stack<Pair<String, MkdirsOptions>> ufsDirsToMakeWithOptions = new Stack<>();
+      AlluxioURI curAlluxioPath = uri.getParent();
+      AlluxioURI curUfsPath = dstPath.getParent();
+      // Stop at the Alluxio root because the mapped directory of Alluxio root in UFS may not exist.
+      while (!mUfs.isDirectory(curUfsPath.toString()) && curAlluxioPath != null) {
+        URIStatus curDirStatus;
+        curDirStatus = fs.getStatus(curAlluxioPath);
+        Permission perm = new Permission(curDirStatus.getOwner(), curDirStatus.getGroup(),
+            (short) curDirStatus.getMode());
+        ufsDirsToMakeWithOptions.push(new Pair<>(curUfsPath.toString(),
+            MkdirsOptions.defaults().setCreateParent(false).setPermission(perm)));
+        curAlluxioPath = curAlluxioPath.getParent();
+        curUfsPath = curUfsPath.getParent();
       }
-      if (numRetry == maxRetry && !mUfs.isDirectory(parentPath)) {
-        throw new IOException(
-            String.format("Failed to create dir: %s after %d retries.", parentPath, numRetry));
+      while (!ufsDirsToMakeWithOptions.empty()) {
+        Pair<String, MkdirsOptions> ufsDirAndPerm = ufsDirsToMakeWithOptions.pop();
+        // UFS mkdirs might fail if the directory is already created. If so, skip the mkdirs
+        // and assume the directory is already prepared, regardless of permission matching.
```

Comment:
File will still be persisted even if the parent permission does not match in UFS and Alluxio. Metadata consistency checker is a periodically service that is used during Alluxio startup or kicked off manually (like FSCK). It will identify any metadata inconsistency between Alluxio namespace and UFS, and fix them by taking the UFS metadata as the source of truth.